File: Examples/quicksort.cpp
#include <stdio.h>

void Print(int *Ar, int size)
{

    for (int k = 0; k < size; k++)
    {
        printf("%d ", Ar[k]);
    }
}

int Partition(int low, int high, int *Ar)
{
    int i = low + 1;
    int j = high;
    int pivot = Ar[low];
    int temp;

    do
    {
        while (Ar[i] <= pivot)
        {
            i++;
        }

        while (Ar[j] >= pivot)
        {
            j--;
        }

        if (i < j)
        {
            temp = Ar[i];
            Ar[i] = Ar[j];
            Ar[j] = temp;
        }
    } while (i < j);

    temp = Ar[low];
    Ar[low] = Ar[j];
    Ar[j] = temp;

    return j;
}

void QuickSort(int *Ar, int low, int high)
{

    int partitionIndex;

    if (low < high)
    {

        partitionIndex = Partition(low, high, Ar);
        QuickSort(Ar, low, partitionIndex - 1);
        QuickSort(Ar, partitionIndex + 1, high);
    }
}

int main()
{

    int Ar[] = {100, 3, 42, 1, 23, 2};
    int size = sizeof(Ar) / sizeof(int);

    printf("Before sort \n");
    Print(Ar, size);

    printf("\nAfter sort \n");
    QuickSort(Ar, 0, size - 1);
    Print(Ar, size);
}

File: Examples/quicksort.py
import random

def quick_sort(array, Lb, Ub):

    if Lb<=Ub:
        pivot = partition(array, Lb, Ub)
        print(pivot)
        quick_sort(array, Lb, pivot-1)
        quick_sort(array, pivot+1, Ub)

def partition(array, Lb, Ub):
    pivot = random.randint(Lb, Ub)
    pivot_element = array[pivot]
    # print('pivot', pivot, 'Ub', Ub, 'Lb', Lb)
    i=Lb
    j=Ub

    while i<j:
        print(i, pivot, 'i pivot')
        while array[i]<=pivot_element and i<Ub:
            i += 1

        # print(j, pivot, 'j pivot')
        while array[j]>=pivot_element and j>Lb:
            j -= 1
        
        if(i<j):
            array[i], array[j] = array[j], array[i]

    array[pivot], array[j] = array[j], array[pivot]

    return j    

array = [8, 7, 2, 1, 0, 9, 6]

quick_sort(array, 0, len(array)-1)

print(array)